### PR TITLE
Fix #672: Add missing fields in ProbeConfig for configuring Readiness/Liveness Probes

### LINKED ISSUE #672: Improve our Liveness/Readiness Probe XML configuration support
Right now our support for [Liveness/Readiness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#http-probes) is okay but I think it's not up to date with latest releases of kubernetes. 

I see some things missing:
- Support for configuring custom headers `httpHeaders`
- Configuring `periodSeconds`
- ~Configuring `terminationGracePeriodSeconds`~ (preview feature) (see https://github.com/eclipse/jkube/issues/720)

### PR DESCRIPTION
**Pull request by rohanKanojia**
## Description
Fix #672 
+ Add missing fields in ProbeConfig for configuring liveness/readiness
+ Added a check to keep host null in case it matches localhost

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [X] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] I have read the [contributing guidelines](https://www.eclipse.org/jkube/contributing)
 - [X] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [X] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [X] I have implemented unit tests to cover my changes
 - [X] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by manusa** [comment]
Please rebase, please fix the code smell

**Comment by rohanKanojia** [comment]
I'm not using IntSupplier for a reason here. It doesn't handle nullable integers well

**Comment by manusa** [comment]
> I'm not using IntSupplier for a reason here. It doesn't handle nullable integers well

Then remove the warning and provide a hint as to why we use Supplier instead:
```java
    @SuppressWarnings("java:S4276") // IntSupplier throws NullPointerException when unboxing null Integers
    private void setTimeoutInProbeIfNotNull(Probe probe, Supplier<Integer> integerSupplier, BiConsumer<Integer, Probe> probeConsumer) {
```

File: jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/handler/ProbeHandler.java [hunk 1]
```diff
@@ -93,6 +89,22 @@ private HTTPGetAction getHTTPGetAction(String getUrl) {
         }
     }
 
+    private String resolveHost(String host) {
+        if (host.equalsIgnoreCase("localhost")) {
+            // No need to set host in case of localhost HTTP request, will default to Pod IP
+            return null;
+        }
+        return host;
+    }
```
Comment (manusa) [inline jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/handler/ProbeHandler.java:98]: Why is this necessary?

Comment (rohanKanojia) [reply jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/handler/ProbeHandler.java:98]: This is done in order to support local probes. If we set the host to null, it is assumed to be pod's IP. This way when a user provides `<getUrl>http://localhost:8080/actuator/health</getUrl>`, it generates a probe like this:
```yaml
        readinessProbe:
          httpGet:
            path: /actuator/health
            port: 8080
            scheme: HTTP
```

Comment (manusa) [reply jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/handler/ProbeHandler.java:98]: I see what it does, but local probes were supported before (weren't they?), so why is this necessary now?

Comment (rohanKanojia) [reply jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/handler/ProbeHandler.java:98]: In order to provide local probes user needed to provide host omitted getUrl like this: `<getUrl>http://:8080/actuator/health</getUrl>` . I thought providing localhost could be a cleaner alternative

Comment (manusa) [reply jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/handler/ProbeHandler.java:98]: We should revert this (it won't consider cases such as 127.0.0.1) and it is misleading for users.
The host part of the [probe config](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#http-probes) should be discouraged
> host: Host name to connect to, defaults to the pod IP. You probably want to set "Host" in httpHeaders instead.

File: kubernetes-maven-plugin/doc/src/main/asciidoc/inc/goals/build/_jkube-resource.adoc [hunk 2]
```diff
@@ -219,7 +219,17 @@ Probe configuration is used for configuring https://kubernetes.io/docs/tasks/con
 | Command to execute for probing.
 
 | `getUrl`
-| Probe this URL.
+| Probe URL for HTTP Probe. Configures HTTP probe fields like `host`, `scheme`, `path` etc by parsing URL. For example, a `<getUrl>http://:8080/health</getUrl>` would result in probe generated with fields set like this:
+
+  host: ""
+
+  path: /actuator/health
```
Comment (manusa) [inline kubernetes-maven-plugin/doc/src/main/asciidoc/inc/goals/build/_jkube-resource.adoc:226]: Shouldn't this be just `/health` ?

File: jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/handler/ProbeHandler.java (hunk 1 above)
Comment (manusa) [reply jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/handler/ProbeHandler.java]: @rohanKanojia , please update the PR description to match the new behavior:
> Added a check to keep host null in case it matches localhost
